pkg/agent/route/route_windows.go

```diff
@@ -205,118 +234,80 @@ func (c *Client) DeleteRoutes(podCIDR *net.IPNet) error {
+	if err := util.ReplaceNetRoute(route); err != nil {
+		return fmt.Errorf("failed to install route for virtual Service IP %s: %w", svcIP.String(), err)
 	}
+	c.serviceRoutes.Store(svcIP.String(), route)
+	klog.InfoS("Added virtual Service IP route", "route", route)
 
 	return nil
 }
 
-// TODO: Follow the code style in Linux that maintains one Service CIDR.
-func (c *Client) addServiceRoute(svcIP net.IP) error {
-	obj, found := c.hostRoutes.Load(svcIP.String())
-	svcIPNet := util.NewIPNet(svcIP)
-
-	// Route: Service IP -> VirtualServiceIPv4 (169.254.0.253)
-	route := &util.Route{
-		LinkIndex:         c.nodeConfig.GatewayConfig.LinkIndex,
-		DestinationSubnet: svcIPNet,
-		GatewayAddress:    config.VirtualServiceIPv4,
-		RouteMetric:       util.MetricHigh,
-	}
-	if found {
-		existingRoute := obj.(*util.Route)
-		if existingRoute.GatewayAddress.Equal(route.GatewayAddress) && existingRoute.RouteMetric == route.RouteMetric {
-			klog.V(2).InfoS("Service route already exists", "DestinationIP", route.DestinationSubnet,
-				"Gateway", route.GatewayAddress, "RouteMetric", route.RouteMetric)
-			return nil
-		}
-		// Remove the existing route if gateway or metric is not as expected.
-		if err := util.RemoveNetRoute(existingRoute); err != nil {
-			return fmt.Errorf("failed to delete existing Service route entry, DestinationIP: %s, Gateway: %s, RouteMetric: %d, err: %v",
-				existingRoute.DestinationSubnet, existingRoute.GatewayAddress, existingRoute.RouteMetric, err)
-		}
-	}
+func (c *Client) addServiceCIDRRoute(serviceCIDR *net.IPNet) error {
+	linkIndex := c.nodeConfig.GatewayConfig.LinkIndex
+	gw := config.VirtualServiceIPv4
+	metric := util.MetricHigh
+	curServiceCIDR := c.serviceIPv4CIDR
 
+	// Generate a route with the new ClusterIP CIDR and install it.
+	route := generateRoute(serviceCIDR, gw, linkIndex, metric)
 	if err := util.ReplaceNetRoute(route); err != nil {
-		return err
+		return fmt.Errorf("failed to install a new Service CIDR route: %w", err)
+	}
+
+	// Store the new ClusterIP CIDR and the new generated route to serviceRoutes. Then the calculated route can be restored
+	// when it was deleted since members of serviceRoutes are synchronized periodically.
+	c.serviceIPv4CIDR = serviceCIDR
+	c.serviceRoutes.Store(clusterIPv4CIDRKey, route)
+
+	if curServiceCIDR != nil {
+		// If current destination CIDR is not nil, the route with current destination CIDR should be uninstalled since
+		// a new route with a newly calculated destination CIDR has been installed.
+		route.DestinationSubnet = curServiceCIDR
+		if err := util.RemoveNetRoute(route); err != nil {
+			if strings.Contains(err.Error(), "No matching MSFT_NetRoute objects") {
+				klog.InfoS("Failed to delete stale ClusterIP route since the route doesn't exist", "route", route)
+			} else {
+				return fmt.Errorf("failed to delete route for for Service CIDR %s: %w", serviceCIDR.String(), err)
+			}
+		}
```

Comment:
Historically, on Linux, if Antrea whose version is before merging PR #3465 (fix #3131), the stale routes are like (assuming that the Service CIDR is 10.96.0.0/12):
- 10.96.0.0/24
- 10.96.0.0/20
- 10.96.0.0/16

Then we need to clean them all. If we don't consider the the version without merging PR #3465, we only need to clean the route generated before restarting.

Reply:
But there are per-IP routes installed for cluster IPs and ingress IPs, how do we clean them on restart?

Reply:
We don't need to clean routes for ingress IPs since they will be always installed again if the Service is not removed or modified during the restarting of Antrea. As for routes for ClusterIPs, I'll add some code to handle this.